### ISSUE TITLE
Noisy label 対抗記~Cassava Leaf Disease Classification~

### ISSUE THREAD

**Issue by GENZITSU** [GENZITSU/UsefulMaterials#48]
## 概要
kaggleのCassava Leaf Disease Classificationで取り組まれたNoisy Labelへの対応方法をまとめた。  
本コンペではtraining data, public LBいずれもNoisy Dataとされ「おみくじコンペ」と称された

### 学習データにfitしすぎない
1. 多くのモデルでのensemble
2. noisy labelに対応できるlossの利用
    1. Label Smoothing
    2. [Bi-Tempered Logistic Loss](https://github.com/mlpanda/bi-tempered-loss-pytorch)
        通常のLoistic Lossが一部の外れ値に大きな影響を受けることを、緩和するためのロス。
        以下の図のように、通常のLogistic LossよりもNoisy Dataに強いことがわかる。
        <img width="600" alt="スクリーンショット 2021-04-17 18 01 28" src="https://user-images.githubusercontent.com/19774756/115108308-aebfe380-9faa-11eb-8478-07502171d6de.png">
        Bi-Tempered Logistic Lossには二つのパラメータがある。
        <img width="600" alt="スクリーンショット 2021-04-17 18 28 49" src="https://user-images.githubusercontent.com/19774756/115108350-f6df0600-9faa-11eb-9d8c-77b30e3e306b.png">
        ちなみに数式は以下。
        <img width="600" alt="スクリーンショット 2021-04-18 0 04 15" src="https://user-images.githubusercontent.com/19774756/115117490-ce223500-9fd9-11eb-8ee0-1789b4b7c172.png">
        <img width="600" alt="スクリーンショット 2021-04-18 0 03 42" src="https://user-images.githubusercontent.com/19774756/115117492-cfebf880-9fd9-11eb-97bb-db8b5cc90ff7.png">

    3. Symmetric Cross Entropy
        <img width="600" alt="スクリーンショット 2021-04-17 18 23 29" src="https://user-images.githubusercontent.com/19774756/115108211-1e819e80-9faa-11eb-91fa-4ec4b280aabd.png">

3. Batch Normalization を学習前半に固定する
<img width="390" alt="スクリーンショット 2021-04-17 18 15 52" src="https://user-images.githubusercontent.com/19774756/115107997-03625f00-9fa9-11eb-906a-7c6dfe8e9328.png">
4. Knowledge Distillation Methodの利用
<img width="600" alt="スクリーンショット 2021-04-17 18 16 20" src="https://user-images.githubusercontent.com/19774756/115108013-14ab6b80-9fa9-11eb-9246-881b556e2183.png">

### Noisy Dataを処理する
1. clean labの利用
<img width="600" alt="スクリーンショット 2021-04-18 0 09 17" src="https://user-images.githubusercontent.com/19774756/115117654-a384ac00-9fda-11eb-90e5-3d101810f183.png">
2. Noise Classifierの作成
    clean labでnoisy dataと判定されたdataから判別機を作成する。
3. 誤り分布の利用
<img width="600" alt="スクリーンショット 2021-04-18 0 15 08" src="https://user-images.githubusercontent.com/19774756/115117814-5f45db80-9fdb-11eb-9c0c-c0254ebaec39.png">
<img width="600" alt="スクリーンショット 2021-04-18 0 15 35" src="https://user-images.githubusercontent.com/19774756/115117819-6240cc00-9fdb-11eb-9078-d91e60b46c58.png">
4. is_noisyフラグ + スタッキング
<img width="845" alt="スクリーンショット 2021-04-18 0 15 54" src="https://user-images.githubusercontent.com/19774756/115117898-c06daf00-9fdb-11eb-9e2a-d3eac5a24f86.png">

## その他テクニック
1. snap_mix
cutmix + cam 
<img width="782" alt="スクリーンショット 2021-04-18 0 23 17" src="https://user-images.githubusercontent.com/19774756/115118036-5bff1f80-9fdc-11eb-9494-efc38c45b622.png">

2. gradient accumulation steps
    複数バッチのgradientをまとめてback propすることで、仮想的に大きなバッチサイズを使った学習を行う。

issue-date: 2018-2-15
https://speakerdeck.com/sansandsoc/diary-against-the-noisy-label
